pkg/auto/sccexporter/auto.go

```diff
@@ -0,0 +1,279 @@
+
+	"github.com/smart-core-os/sc-api/go/traits"
+	"github.com/smart-core-os/sc-bos/pkg/auto"
+	"github.com/smart-core-os/sc-bos/pkg/auto/sccexporter/config"
+	"github.com/smart-core-os/sc-bos/pkg/gen"
+	meterpb "github.com/smart-core-os/sc-bos/pkg/gentrait/meter"
+	"github.com/smart-core-os/sc-bos/pkg/node"
+	"github.com/smart-core-os/sc-bos/pkg/task/service"
+	"github.com/smart-core-os/sc-golang/pkg/trait"
+)
+
+const AutoName = "sccexporter"
+
+var Factory auto.Factory = factory{}
+
+type factory struct{}
+
+type AutoImpl struct {
+	*service.Service[config.Root]
+	auto.Services
+
+	airQualityClient     traits.AirQualitySensorApiClient
+	airTemperatureClient traits.AirTemperatureApiClient
+	metadataClient       traits.MetadataApiClient
+	meterClient          gen.MeterApiClient
+	meterInfoClient      gen.MeterInfoClient
+	occupancyClient      traits.OccupancySensorApiClient
+}
+
+func (f factory) New(services auto.Services) service.Lifecycle {
+	a := &AutoImpl{
+		Services: services,
+	}
+	a.Service = service.New(service.MonoApply(a.applyConfig), service.WithParser(config.ParseConfig))
+	a.Logger = a.Logger.Named(AutoName)
+	return a
+}
+
+func (a *AutoImpl) initialiseClients(n *node.Node) {
+	a.airQualityClient = traits.NewAirQualitySensorApiClient(n.ClientConn())
+	a.airTemperatureClient = traits.NewAirTemperatureApiClient(n.ClientConn())
+	a.metadataClient = traits.NewMetadataApiClient(n.ClientConn())
+	a.meterClient = gen.NewMeterApiClient(n.ClientConn())
+	a.meterInfoClient = gen.NewMeterInfoClient(n.ClientConn())
+	a.occupancyClient = traits.NewOccupancySensorApiClient(n.ClientConn())
+}
+
+func (a *AutoImpl) applyConfig(ctx context.Context, cfg config.Root) error {
+
+	a.initialiseClients(a.Node)
+
+	grp, autoCtx := errgroup.WithContext(ctx)
+	allDevices := make(map[string]*device)
+
+	// discover all the devices which implement the configured traits and set up the allDevices map
+	for _, traitName := range cfg.Traits {
+		err := a.getAllTraitImplementors(autoCtx, trait.Name(traitName), allDevices)
+		if err != nil {
+			a.Logger.Error("failed to get devices for trait", zap.String("trait", traitName), zap.Error(err))
+			return err
+		}
+
+		switch traitName {
+		case string(meterpb.TraitName):
+			// grab the trait info for all meters first and save it in the device so we can push it
+			// only supports the Meter info, think it's the only one we really need for data...
+			a.getMeterInfo(autoCtx, trait.Name(traitName), allDevices)
+		}
+	}
+
+	mqttClient, err := newMqttClient(cfg.Mqtt)
+	if err != nil {
+		a.Logger.Error("failed to create mqtt client", zap.Error(err))
+		return err
+	}
+
+	s := newSccConnector(a.Logger, cfg.Mqtt, mqttClient)
+
+	grp.Go(func() error {
+		return s.publishToScc(autoCtx)
+	})
+
+	t := time.Now()
+	iterationCount := 0
+	grp.Go(func() error {
+		for {
+			next := cfg.Mqtt.SendInterval.Next(t)
+			select {
+			case <-autoCtx.Done():
+				return nil
+			case <-time.After(time.Until(next)):
+				t = next
```

Comment:
I think technically this should be `time.Now()` or even the returned value from the timer chan